transport-native-io_uring/src/main/c/netty_io_uring_native.c

```diff
@@ -0,0 +1,525 @@
+  if (!ret) {
+    io_uring_ring->flags = p->flags;
+    io_uring_ring->ring_fd = ring_fd;
+  } else {
+    perror("setup_io_uring error \n");
+  }
+}
+
+void io_uring_prep_write(struct io_uring_sqe *sqe, int fd, const void *buf,
+                         unsigned nbytes, off_t offset) {
+  io_uring_prep_rw(IORING_OP_WRITE, sqe, fd, buf, nbytes, offset);
+}
+
+void io_uring_prep_read(struct io_uring_sqe *sqe, int fd, void *buf,
+                        unsigned nbytes, off_t offset) {
+  io_uring_prep_rw(IORING_OP_READ, sqe, fd, buf, nbytes, offset);
+}
+
+void io_uring_sqe_set_data(struct io_uring_sqe *sqe, unsigned long data) {
+  sqe->user_data = (unsigned long)data;
+}
+
+void queue_read(int file_fd, struct io_uring *ring, void *buffer, jint event_id,
+                jint pos, jint limit) {
+  struct io_uring_sqe *sqe = NULL;
+
+  sqe = io_uring_get_sqe(ring);
+  if (!sqe) {
+    fprintf(stderr, "Could not get SQE.\n");
+    return;
+  }
+  io_uring_prep_read(sqe, file_fd, buffer + pos, (size_t)(limit - pos), 0);
+  io_uring_sqe_set_data(sqe, (int)event_id);
+}
+
+void queue_write(int file_fd, struct io_uring *ring, void *buffer,
+                 jint event_id, jint pos, jint limit) {
+  struct io_uring_sqe *sqe;
+
+  sqe = io_uring_get_sqe(ring);
+  if (!sqe) {
+    fprintf(stderr, "Could not get SQE.\n");
+    return;
+  }
+  io_uring_prep_write(sqe, file_fd, buffer + pos, (size_t)(limit - pos), 0);
+  io_uring_sqe_set_data(sqe, (unsigned long)event_id);
+}
+
+int __io_uring_peek_cqe(struct io_uring *ring, struct io_uring_cqe **cqe_ptr) {
+  struct io_uring_cqe *cqe;
+  unsigned head;
+  int err = 0;
+
+  do {
+    io_uring_for_each_cqe(ring, head, cqe) break;
+    break;
+  } while (1);
+
+  *cqe_ptr = cqe;
+  return err;
+}
+
+long io_uring_wait_cqe(struct io_uring *ring, unsigned wait_nr) {
+  struct io_uring_cqe *cqe = NULL;
+  int ret = 0, err;
+  unsigned flags = 0;
+
+  err = __io_uring_peek_cqe(ring, &cqe);
+  if (err) {
+    printf("error peek \n");
+    return -errno;
+  }
+
+  if (cqe) {
+    return (long)cqe;
+  }
+
+  flags = IORING_ENTER_GETEVENTS;
+  ret = sys_io_uring_enter(ring->ring_fd, 0, wait_nr, flags, NULL);
+
+  if (ret < 0) {
+    return -1;
+  } else if (ret == 0) {
+
+    err = __io_uring_peek_cqe(ring, &cqe);
+    if (err) {
+      printf("error peek \n");
+      return -1;
+    }
+    if (cqe) {
+      return (long)cqe;
+    }
+  }
+  return -1;
+}
+
+/*
+ * Submit sqes acquired from io_uring_get_sqe() to the kernel.
+ *
+ * Returns number of sqes submitted
+ */
+int io_uring_submit(struct io_uring *ring) {
+  int submitted = io_uring_flush_sq(ring);
+  int ret;
+
+  ret = sys_io_uring_enter(ring->ring_fd, submitted, 0, 0, NULL);
+  if (ret < 0)
+    return -errno;
+  return ret;
+}
+
+// all jni methods
+
+static jlong netty_io_uring_setup(JNIEnv *env, jclass class1, jint entries) {
+  struct io_uring_params p;
+  memset(&p, 0, sizeof(p));
+
+  int ring_fd = sys_io_uring_setup((int)entries, &p);
```

Comment:
We need to capture the error condition when the `ring_fd` is `-1`. We should especially watch out for `ENOSYS`.